server/komodo-rest/src/main/java/org/komodo/rest/service/KomodoMetadataService.java

```diff
@@ -3576,17 +3581,97 @@ public Response publishVirtualization(final @Context HttpHeaders headers, final
         }
         UnitOfWork uow = null;
         try {
-            Repository repo = this.kengine.getDefaultRepository();
-            uow = createTransaction(principal, "publish", true); //$NON-NLS-1$
+            uow = createTransaction(principal, "publish-init", true); //$NON-NLS-1$
             Vdb vdb = findVdb(uow, attributes.getName());
+            if (vdb == null) {
+                //
+                // We don't have a vdb so maybe we have a dataservice instead
+                // Find the dataservice's vdb to publish.
+                //
+                Dataservice dataservice = findDataservice(uow, attributes.getName());
```

Comment:
What is a DataService in this context of the Komodo application? Are there different ways of VDB availability?

Reply:
Beetle studio, like DSB, interacts with teiid-komodo at the 'dataservice' layer (you may recall I started a discussion on IRC a week or so ago concering the need for dataservices). Therefore, the vdb created by beetle-studio/teiid-komodo when the user has completed the GUI wizard is in fact wrapped in a dataservice. Now, it is possible to extract the 'service-vdb' (the main vdb of the dataservice since the latter can contain others for import), however, making beetle-studio do that work is unnecessary given that the dataservice API remains.

At some point, this can be removed along with the entire dataservice framework if that is felt to be the way forward but for now that is out of scope for this sprint. Therefore, allowing this method to take the name of a vdb or a dataservice covers all options and reduces the amount of business logic necessary in beetle-studio.

Reply:
Unfortunately, I was not part of the IRC discussion. I see your point. I strongly encourage the removal of Data Service to bring clarity to API